scraper/node_data.py

```diff
@@ -0,0 +1,209 @@
+    """
+
+    def __init__(self, nodes_filename='nodes.json', urls_filename='urls.txt', encoding='latin-1'):
+        """
+        Set some defaults for lazy loading private storage and defining where
+        the files live.
+
+        :param str nodes_filename: Path and filename of the nodes.json file.
+        :param str urls_filename: Path and filename of the urls.txt file.
+        :param str encoding: File encoding.
+        """
+        self._data = None
+        self._urls = None
+        self.nodes_filename = nodes_filename
+        self.urls_filename = urls_filename
+        self.encoding = encoding
+
+    def get_node_data_by_url(self, url):
+        """
+        Get the data for a single node given a url.
+
+        :param str url: The full url of the node.
+        :return: The data about the node.
+        :rtype: dict
+        """
+        return self.get_data()[url]
+
+    def get_data(self):
+        """
+        Get the node data. Calls the subclass's private method and caches it.
+
+        :return: The node data, keyed by the node's full url.
+        :rtype: dict
+        """
+        if self._data is None:
+            self._data = self._query_data()
+
+        return self._data
+
+    def get_urls(self):
+        """
+        Get the urls of stories to provide to the spider. Calls the subclass's
+        private method and caches it.
+
+        :return: A list of urls.
+        :rtype: list
+        """
+        if self._urls is None:
+            self._urls = self._query_urls()
+
+        return self._urls
+
+    def _query_data(self):
+        """
+        Implement this in the subclass to return the dict of node data keyed
+        by url.
+        """
+        return {}
+
+    def _query_urls(self):
+        """
+        Implement this in the subclass to return the list of urls to stories.
+        """
+        return []
+
+
+class Reader(AbstractNodeData):
+    """
+    The node data Reader. Opens the files and provides the node data and urls.
+    """
+
+    def _query_data(self):
+        """
+        Retrieve the node data from the nodes.json file.
+        """
+        with codecs.open(self.nodes_filename, 'rb') as fp:
+            all_node_data = json.load(fp, encoding=self.encoding)
+
+        return all_node_data
+
+    def _query_urls(self):
+        """
+        Retrieves the urls from urlst.txt.
+        """
+        with codecs.open(self.urls_filename, 'rb', encoding=self.encoding) as fp:
+            urls = fp.readlines()
+
+        return [x.strip() for x in urls]
+
+
+class Writer(AbstractNodeData):
+    """
+    The node data writer. This queries the database and writes the files.
+    """
+
+    def write_nodes_json(self):
+        """
+        Write the nodes.json file.
+        """
+        with codecs.open(self.nodes_filename, 'wb') as fp:
+            json.dump(self.get_data(), fp, encoding=self.encoding)
+
+    def write_urls_txt_file(self):
```

Comment:
This is where we write the urls.txt file after getting the database results.

encoding is latin-1.